File: Python/hackerrank/common-child.py
#!/bin/python3

import math
import os
import random
import re
import sys

import numpy as np

'''
Common Child

https://www.hackerrank.com/challenges/common-child/problem
'''
# Complete the commonChild function below.
def commonChild(s1, s2):
    len1 = len(s1)
    len2 = len(s2)
    c =[[ 0 for _ in range(len2+1)] for _ in range(len1+1)]
    for i in range(len1):
        for j in range(len2):
            if s1[i] == s2[j]:
                c[i + 1][j + 1] = c[i][j] + 1
            elif c[i + 1][j] > c[i][j + 1]:
                c[i + 1][j + 1] = c[i + 1][j]
            else:
                c[i + 1][j + 1] = c[i][j + 1]
        #print(c[i])
    return c[len1][len2]


def printLcs(flag, a, i, j):
    if i == 0 or j == 0:
        return
    if flag[i][j] == 'ok':
        printLcs(flag, a, i - 1, j - 1)
        print(a[i - 1], end='')
    elif flag[i][j] == 'left':
        printLcs(flag, a, i, j - 1)
    else:
        printLcs(flag, a, i - 1, j)
    
    

if __name__ == '__main__':

    s1 = "SHINCHAN"

    s2 = "NOHARAAA"
    
    result = commonChild(s1, s2)
    print(result)

File: Go/common_child.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"strings"
)

/*
Common Child

https://www.hackerrank.com/challenges/common-child/problem
*/

// Complete the commonChild function below.

func commonChild(s1 string, s2 string) int32 {
	length := len(s2)
	matrix := make([][]int32, length+1)
	//matrix[length][length] = 0
	for i := 0; i <= length; i++ {
		matrix[i] = make([]int32, length+1)
	}
	a := []byte(s1)
	b := []byte(s2)
	for i := 0; i < length; i++ {
		for j := 0; j < length; j++ {
			if a[i] == b[j] {
				matrix[i+1][j+1] = matrix[i][j] + 1
			} else {
				matrix[i+1][j+1] = max(matrix[i+1][j], matrix[i][j+1])
			}
		}
		//fmt.Printf("%#v\n", matrix[i])
	}
	return matrix[length][length]
}
func max(a, b int32) int32 {
	if a > b {
		return a
	}
	return b

}

func common() {

	s1 := "SHINCHAN"

	s2 := "NOHARAAA"

	result := commonChild(s1, s2)

	fmt.Printf("result=%d\n", result)

}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
